### DISCUSSION TITLE
Is there support for readonly arrays?

### DISCUSSION THREAD

**Discussion by thes01** [sinclairzx81/typebox#860] [Q&A] [answered]
Typescript has a special type for immutable arrays (excerpt from the [official docs](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#readonly-and-const))

```ts
let a: ReadonlyArray<number> = [1, 2, 3];
let b: readonly number[] = [1, 2, 3];
a.push(102); // error
b[0] = 101; // error
```

Is there any support for such type to be expressed in Typebox schemas?
I could only find `Type.Readonly`, which only works for object properties. Currently, I use `Type.Unsafe` to cast the type to a correct one.

Thank you!

**Comment by sinclairzx81** [comment, marked as answer]
@thes01 Hi, 

> Is there any support for such type to be expressed in Typebox schemas?

Unfortunately, there's not currently a way to infer isolated readonly tuples (i.e. those not embedded in outer types) but Unsafe is a reasonable workaround. The reason has to do with the way TB inference works; where an outer type (TObject, TFunction and TConstructor) will ascribe the `readonly` modifier to the interior type. As isolated types don't have an outer type, there is nothing to ascribe the modifier.

This said, following is a ReadonlyTuple type that will perform the isolated readonly inference via Unsafe (if you need a a type to handle this)

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxjsriQBYYgFwC+cAZlBCHAOQACAzsADsSAGwLAoAehjoMAIwgAPTgChlGBZFhxpmFACUMBACYQBw1MgCuYYRgA8yOOpgYBR3nCiGTZ1Cko0dADaALo4AIIkJE4KLm4eXsam5nCWAgDWAhAA7gKhcAC8cKEAfIXKcCgxce7FgswYUHAAMtWutRTUtAQ4AHT99Y1wem3x-l3BISEVlXAA-PreyRbWtnZ6OIk+KUH9vZFkeITEJHbNJSElM5UAXHAHqiSmvPAGSb5WNhiFcA6jHQFuqEcCMisg3tsVl8HCUSgAKHQYXh3Xb9ZAhACUhSulTQmF6AFUBLwCA11vC8Rhep9bAiZLwMRjVMonsT4I4ihDljSMHCgpTevgoIIAOZwjHkGS9ABylhAska4sxqkRVSK+CIpDsiIgzBQZQkEk8S18xRewoEIpwAjlCqgISAA)

```typescript
import { Type, Static, TSchema } from '@sinclair/typebox'

export type TReadonlyTuple<T extends readonly TSchema[], Acc extends readonly unknown[] = []> =
  T extends [infer L extends TSchema, ...infer R extends TSchema[]]
    ? TReadonlyTuple<R, readonly [...Acc, Static<L>]>
    : Acc

const ReadonlyTuple = <T extends TSchema[], R = TReadonlyTuple<T>>(types: [...T]) =>
  Type.Unsafe<R>(Type.Tuple(types))


const T = ReadonlyTuple([Type.String(), Type.Number()])

type T = Static<typeof T> // readonly [string, number]
```
I intend to revisit this aspect of TypeBox when it drops support for TS 4.x as there is functionality specific to 5.x that may assist here. Initial updates will likely apply specifically to Type.Const (short term) with isolated readonly more broadly implemented in later revisions.

Hope this helps
S

**Reply by thes01** [reply]
Thank you very much for the clarification and example.
